packages/forms/src/UIForm/fieldsets/Array/Array.component.js

```diff
@@ -0,0 +1,192 @@
+import React, { PropTypes } from 'react';
+import ArrayItem from './ArrayItem.component';
+import Message from '../../Message';
+import Widget from '../../Widget';
+import { shiftArrayErrorsKeys } from '../../utils/validation';
+
+import theme from './Array.scss';
+
+function adaptKeyWithIndex(keys, index) {
+	let indexedKeys = keys;
+	const firstIndexPlaceholder = indexedKeys.indexOf('');
+	if (firstIndexPlaceholder >= 0) {
+		indexedKeys = [...keys];
+		indexedKeys[firstIndexPlaceholder] = index;
+	}
+	return indexedKeys;
+}
+
+function getRange(previousIndex, nextIndex) {
+	if (previousIndex < nextIndex) {
+		return {
+			minIndex: previousIndex,
+			maxIndex: nextIndex + 1,
+		};
+	}
+
+	return {
+		minIndex: nextIndex,
+		maxIndex: previousIndex + 1,
+	};
+}
+
+function getItemSchema(arraySchema, index) {
+	// insert index in all fields
+	const items = arraySchema.items.map(item => ({
+		...item,
+		key: adaptKeyWithIndex(item.key, index),
+	}));
+
+	// insert index in item schema key
+	const key = arraySchema.key.concat(index);
+
+	return {
+		key,
+		items,
+		widget: arraySchema.itemWidget || 'fieldset',
+	};
+}
+
+export default class ArrayWidget extends React.Component {
+	constructor(props) {
+		super(props);
+
+		this.onAdd = this.onAdd.bind(this);
+		this.onRemove = this.onRemove.bind(this);
+		this.onReorder = this.onReorder.bind(this);
+	}
+
+	onAdd(event) {
+		const arrayMergedSchema = this.props.schema;
+		const defaultValue = arrayMergedSchema.schema.items.type === 'object' ? {} : '';
+		const value = [...this.props.value, defaultValue];
+
+		this.props.onChange(
+			event,
+			{ schema: arrayMergedSchema, value }
+		);
+	}
+
+	onRemove(event, indexToRemove) {
+		const schema = this.props.schema;
+		const value = this.props.value.slice(0);
+		value.splice(indexToRemove, 1);
+
+		return this.props.onChange(
+			event,
+			{ schema, value },
+			{
+				widgetChangeErrors(errors) {
```

Comment:
I think this could be clearer if `widgetChangeErrors` were defined before the `return`